xbmc/addons/interfaces/GUI/Window.cpp

```diff
@@ -0,0 +1,657 @@
+  pAddonWindow->CBGetContextButtons = CBGetContextButtons;
+  pAddonWindow->CBOnContextButton = CBOnContextButton;
+  Interface_GUIGeneral::unlock();
+}
+
+bool Interface_GUIWindow::show(void* kodiBase, void* handle)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (!addon)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIWindow::%s - invalid data", __FUNCTION__);
+    return false;
+  }
+
+  CGUIAddonWindow *pAddonWindow = static_cast<CGUIAddonWindow*>(handle);
+  if (!pAddonWindow)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIWindow::%s - No Window on addon '%s'", __FUNCTION__, addon->ID().c_str());
+    return false;
+  }
+
+  if (pAddonWindow->m_oldWindowId != pAddonWindow->m_windowId && 
+      pAddonWindow->m_windowId != g_windowManager.GetActiveWindow())
+    pAddonWindow->m_oldWindowId = g_windowManager.GetActiveWindow();
+
+  Interface_GUIGeneral::lock();
+  if (pAddonWindow->IsDialog())
+    dynamic_cast<CGUIAddonWindowDialog*>(pAddonWindow)->Show();
+  else
+    g_windowManager.ActivateWindow(pAddonWindow->GetID());
+  Interface_GUIGeneral::unlock();
+  
+  return true;
+}
+
+bool Interface_GUIWindow::close(void* kodiBase, void* handle)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (!addon)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIWindow::%s - invalid data", __FUNCTION__);
+    return false;
+  }
+
+  CGUIAddonWindow *pAddonWindow = static_cast<CGUIAddonWindow*>(handle);
+  if (!pAddonWindow)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIWindow::%s - No Window on addon '%s'", __FUNCTION__, addon->ID().c_str());
+    return false;
+  }
+
+  pAddonWindow->m_bModal = false;
+  if (pAddonWindow->IsDialog())
+    dynamic_cast<CGUIAddonWindowDialog*>(pAddonWindow)->PulseActionEvent();
+  else
+    dynamic_cast<CGUIAddonWindow*>(pAddonWindow)->PulseActionEvent();
+
+  Interface_GUIGeneral::lock();
+
+  // if it's a dialog, we have to close it a bit different
+  if (pAddonWindow->IsDialog())
+    dynamic_cast<CGUIAddonWindowDialog*>(pAddonWindow)->Show(false);
+  else
+    g_windowManager.ActivateWindow(pAddonWindow->m_oldWindowId);
+  pAddonWindow->m_oldWindowId = 0;
+
+  Interface_GUIGeneral::unlock();
+
+  return true;
+}
+
+bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (!addon)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIWindow::%s - invalid data", __FUNCTION__);
+    return false;
+  }
+
+  CGUIAddonWindow *pAddonWindow = static_cast<CGUIAddonWindow*>(handle);
+  if (!pAddonWindow)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIWindow::%s - No Window on addon '%s'", __FUNCTION__, addon->ID().c_str());
+    return false;
+  }
+
+  pAddonWindow->m_bModal = true;
+
+  if (pAddonWindow->GetID() != g_windowManager.GetActiveWindow())
+    show(kodiBase, handle);
+
+  return true;
+}
+
+int Interface_GUIWindow::GetNextAvailableWindowId()
+{
+  Interface_GUIGeneral::lock();
+
+  // window id's 14000 - 14100 are reserved for addons
```

Comment:
I would remove the hard numbers here and in the following lines as we have WINDOW_ADDON_START and WINDOW_ADDON_END